### ISSUE TITLE
Triton Error: OutOfResources Shared Memory

### ISSUE THREAD

**Issue by jsie7** [state-spaces/mamba#387]
I'm trying to rerun the MQAR experiments in Figure 8 of the paper. For model dimensions <= 256 this works fine. However, when I increase model size to 512, I always run into the following error:
`triton.runtime.autotuner.OutOfResources: out of resource: shared memory, Required: 663808, Hardware limit: 101376. Reducing block sizes or `num_stages` may help.`
The error persists even when reducing the `chunk_size`, increasing GPU memory, or using a different GPU (RTX 3090, RTX 4090, A100; the Hardware limit in above error obviously changes between these, but it's never enough).
It's weird that it complains about shared memory, which is not a lack of VRAM (GPU memory) but essentially conventional RAM. I'm no expert in trition so I don't really know why the autotuner evne tries to allocate shared memory or why it needs so much.
Did anybody also encounter this problem and was able to fix it?

**Comment by tridao** [comment]
Can you give a short script to reproduce the error?

**Comment by jsie7** [comment]
**TL;DR**: The issue is with the head dimension. It seems that `headdim = 512` does not work for any GPU due to how the kernel stores the parameters in memory.

ok I think I figured it out. The issue is not with the model dimension but with the head dimension, since I set `d_inner == headdim` to only get one head. The error
`triton.runtime.autotuner.OutOfResources: out of resource: shared memory, Required: 663808, Hardware limit: 101376.`
occurs irrespective of any other dimension and only depends on the head dimension. I assume this is due to how the parameters are aligned in memory (but I'm really no expert here).
I found that `headdim = 256` works for A100 (40/80 GB) but not RTX 3090/4090 and that `headdim = 512` doesn't work, even for A100. Attached is a simple implementation to test this (basically a stripped down version of MQAR in [zoology](https://github.com/HazyResearch/zoology)). To run adapt `D_MODEL, HEADDIM` at the start of `train_simple.py` and then run it. However, note that the script uses `expand = 1` which throws the error discussed in #351. Hence it uses the slow path and PyTorch Conv1d (which throws another error due to a bug; I opened a separate issue #393). Thus, the code might not work out of the box.

Code: [train_MQAR.zip](https://github.com/user-attachments/files/15816687/train_MQAR.zip)

**Comment by tridao** [comment]
Yeah that makes sense. We only ever run with headdim <= 256.

**Comment by chain2k** [comment]
Installed torch 2.7 and zluda 3.9.3. Got an error with triton 3.3.0. I'll be back home in May, I'll be able to write down which one. Installed triton 3.0.0, everything seems to work fine there,  but there is a different error:
`triton.runtime.errors.OutOfResources: out of resource: shared memory, Required: 106496, Hardware limit: 65536. Reducing block sizes or `num_stages` may help.`
Does anyone know how to lower these values? Or I found this solution, is there any way to customize it?
[https://github.com/state-spaces/mamba/issues/387](url)

**Comment by hanxiaomeng-w** [comment]
> **TL;DR**：问题出在头部尺寸上。`headdim = 512`由于内核在内存中存储参数的方式，这似乎不适用于任何 GPU。
> 
> 好的，我想我明白了。问题不在于模型尺寸，而在于头部尺寸，因为我设置`d_inner == headdim`为只获取一个头部。无论其他任何尺寸如何，都会发生错误 `triton.runtime.autotuner.OutOfResources: out of resource: shared memory, Required: 663808, Hardware limit: 101376.` ，并且仅取决于头部尺寸。我认为这是由于参数在内存中的对齐方式造成的（但我真的不是这方面的专家）。 我发现这`headdim = 256`适用于 A100（40/80 GB），但不适用于 RTX 3090/4090，而且`headdim = 512`即使对于 A100 也不起作用。附件是一个简单的实现来测试这个（基本上是[动物学](https://github.com/HazyResearch/zoology)`D_MODEL, HEADDIM`中 MQAR 的精简版）。在开始时运行 adapt`train_simple.py`然后运行它。但是，请注意，脚本使用会引发[#351](https://github.com/state-spaces/mamba/issues/351)`expand = 1`中讨论的错误。因此，它使用慢速路径和 PyTorch Conv1d（由于错误而引发另一个错误；我打开了一个单独的问题[#393](https://github.com/state-spaces/mamba/issues/393)）。因此，代码可能无法开箱即用。[](https://github.com/state-spaces/mamba/issues/351)[](https://github.com/state-spaces/mamba/issues/393)
> 
> 代码：[train_MQAR.zip](https://github.com/user-attachments/files/15816687/train_MQAR.zip)

谢谢，把4090换成a800便解决了
